modules/apps/transfer/keeper/forwarding.go

```diff
@@ -97,3 +96,62 @@ func (k Keeper) revertForwardedPacket(ctx sdk.Context, prevPacket channeltypes.P
 	}
 	return nil
 }
+
+// forwardPacket forwards a fungible FungibleTokenPacketDataV2 to the next hop in the forwarding path.
+func (k Keeper) forwardPacket(ctx sdk.Context, data types.FungibleTokenPacketDataV2, packet channeltypes.Packet, receivedCoins sdk.Coins) error {
+	var memo string
+
+	var nextForwardingPath *types.Forwarding
+	if len(data.Forwarding.Hops) == 1 {
+		memo = data.Forwarding.Memo
+		nextForwardingPath = nil
+	} else {
+		nextForwardingPath = &types.Forwarding{
+			Hops: data.Forwarding.Hops[1:],
+			Memo: data.Forwarding.Memo,
+		}
+	}
+
+	// sending from the forward escrow address to the original receiver address.
+	sender := types.GetForwardAddress(packet.DestinationPort, packet.DestinationChannel)
+
+	msg := types.NewMsgTransfer(
+		data.Forwarding.Hops[0].PortId,
+		data.Forwarding.Hops[0].ChannelId,
+		receivedCoins,
+		sender.String(),
+		data.Receiver,
+		packet.TimeoutHeight,
+		packet.TimeoutTimestamp,
+		memo,
+		nextForwardingPath,
+	)
+
+	resp, err := k.Transfer(ctx, msg)
+	if err != nil {
+		return err
+	}
+
+	k.SetForwardedPacket(ctx, data.Forwarding.Hops[0].PortId, data.Forwarding.Hops[0].ChannelId, resp.Sequence, packet)
+	return nil
+}
+
+// getReceiverFromPacketData returns either the sender specified in the packet data or the forwarding address
+// if there are still hops left to perform.
+func getReceiverFromPacketData(data types.FungibleTokenPacketDataV2, portID, channelID string) (sdk.AccAddress, error) {
+	receiver, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return nil, errorsmod.Wrapf(ibcerrors.ErrInvalidAddress, "failed to decode receiver address %s: %v", data.Receiver, err)
+	}
+
+	if shouldForwardPacketData(data) {
+		receiver = types.GetForwardAddress(portID, channelID)
+	}
+
+	return receiver, nil
+}
+
+// shouldForwardPacketData determines if the packet should be forwarded to the next hop.
+func shouldForwardPacketData(data types.FungibleTokenPacketDataV2) bool {
```

Comment:
I have the exact same check in the changes for authz. Should this be added as a convenience method on FungibleTokenPacketDataV2 directly?

Reply:
not a bad idea :+1: I'd be happy enough to do that